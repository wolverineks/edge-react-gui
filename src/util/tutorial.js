// @flow
import type { Disklet } from 'disklet'

const TUTORIAL = 'tutorial.json'

type UserTutorialList = {
  walletListSlideTutorialCount: number
}

export const getWalletListSlideTutorial = async (userId: string, disklet: Disklet): Promise<UserTutorialList> => {
  try {
    const userTutorialList = JSON.parse(await disklet.getText(TUTORIAL))
    return userTutorialList
  } catch (error) {
    console.log(error)
    return { walletListSlideTutorialCount: 0 }
  }
}

export const setUserTutorialList = async (data: UserTutorialList, disklet: Disklet): Promise<void> => {
  await disklet.setText(TUTORIAL, JSON.stringify(data))
}
