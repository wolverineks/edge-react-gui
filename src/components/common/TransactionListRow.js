// @flow

import { bns } from 'biggystring'
import type { EdgeCurrencyInfo, EdgeCurrencyWallet } from 'edge-core-js'
import * as React from 'react'
import { Image, TouchableOpacity, View } from 'react-native'
import { Actions } from 'react-native-router-flux'
import { connect } from 'react-redux'
import { sprintf } from 'sprintf-js'

import { Fontello } from '../../assets/vector/index.js'
import * as intl from '../../locales/intl.js'
import s from '../../locales/strings'
import { type RootState } from '../../types/reduxTypes.js'
import type { TransactionListTx } from '../../types/types.js'
import * as UTILS from '../../util/utils'
import { getDenomination, getFiatSymbol } from '../../util/utils.js'
import { showError } from '../services/AirshipInstance.js'
import { type Theme, type ThemeProps, cacheStyles, withTheme } from '../services/ThemeContext.js'
import { EdgeText } from '../themed/EdgeText.js'

type StateProps = {
  cryptoAmount: string,
  denominationSymbol?: string,
  fiatAmount: string,
  fiatSymbol: string,
  isSentTransaction: boolean,
  requiredConfirmations: number,
  selectedCurrencyName: string,
  thumbnailPath?: string,
  walletBlockHeight: number
}

type OwnProps = {
  walletId: string,
  currencyCode: string,
  transaction: TransactionListTx
}

type Props = OwnProps & StateProps & ThemeProps

export class TransactionListRowComponent extends React.PureComponent<Props> {
  handlePress = () => {
    const { transaction, thumbnailPath } = this.props
    if (transaction) {
      Actions.transactionDetails({ edgeTransaction: transaction, thumbnailPath })
    } else {
      showError(s.strings.transaction_details_error_invalid)
    }
  }

  render() {
    // What is this for?
    global.pcount && global.pcount('TransactionRow:render')

    const {
      cryptoAmount,
      denominationSymbol,
      fiatAmount,
      fiatSymbol,
      isSentTransaction,
      requiredConfirmations,
      selectedCurrencyName,
      theme,
      thumbnailPath,
      transaction,
      walletBlockHeight
    } = this.props
    const styles = getStyles(theme)

    const cryptoAmountString = `${isSentTransaction ? '-' : '+'} ${denominationSymbol ? denominationSymbol + ' ' : ''}${cryptoAmount}`
    const fiatAmountString = `${fiatSymbol} ${fiatAmount}`

    // Transaction Text and Icon
    let transactionText, transactionIcon, transactionStyle
    if (isSentTransaction) {
      transactionText =
        transaction.metadata && transaction.metadata.name ? transaction.metadata.name : s.strings.fragment_transaction_list_sent_prefix + selectedCurrencyName
      transactionIcon = <Fontello name="send" size={theme.rem(1.75)} color={theme.negativeText} />
      transactionStyle = styles.iconSent
    } else {
      transactionText =
        transaction.metadata && transaction.metadata.name
          ? transaction.metadata.name
          : s.strings.fragment_transaction_list_receive_prefix + selectedCurrencyName
      transactionIcon = <Fontello name="request" size={theme.rem(1.75)} color={theme.positiveText} />
      transactionStyle = styles.iconRequest
    }

    // Pending Text and Style
    const currentConfirmations = walletBlockHeight && transaction.blockHeight > 0 ? walletBlockHeight - transaction.blockHeight + 1 : 0
    let pendingText, pendingStyle
    if (walletBlockHeight === 0) {
      pendingText = s.strings.fragment_transaction_list_tx_synchronizing
      pendingStyle = styles.partialTime
    } else if (transaction.blockHeight < 0) {
      pendingText = s.strings.fragment_transaction_list_tx_dropped
      pendingStyle = styles.partialTime
    } else if (currentConfirmations <= 0) {
      // if completely unconfirmed or wallet uninitialized, or wallet lagging behind (transaction block height larger than wallet block height)
      pendingText = s.strings.fragment_wallet_unconfirmed
      pendingStyle = styles.pendingTime
    } else if (currentConfirmations < requiredConfirmations) {
      pendingText = sprintf(s.strings.fragment_transaction_list_confirmation_progress, currentConfirmations, requiredConfirmations)
      pendingStyle = styles.partialTime
    } else {
      pendingText = transaction.time
      pendingStyle = styles.completedTime
    }

    // Transaction Category
    let categoryText
    const transactionCategory = transaction.metadata ? transaction.metadata.category : null
    if (transactionCategory) {
      const splittedFullCategory = UTILS.splitTransactionCategory(transactionCategory)
      const { category, subCategory } = splittedFullCategory
      if (subCategory) {
        const mainCategory = category.toLowerCase()
        switch (mainCategory) {
          case 'exchange':
            categoryText = `${s.strings.fragment_transaction_exchange}:${subCategory}`
            break
          case 'expense':
            categoryText = `${s.strings.fragment_transaction_expense}:${subCategory}`
            break
          case 'transfer':
            categoryText = `${s.strings.fragment_transaction_transfer}:${subCategory}`
            break
          case 'income':
            categoryText = `${s.strings.fragment_transaction_income}:${subCategory}`
            break
          default:
            break
        }
      }
    }

    return (
      <TouchableOpacity onPress={this.handlePress}>
        <View style={styles.rowContainer}>
          <View style={[styles.iconContainer, transactionStyle]}>
            {thumbnailPath ? <Image style={styles.icon} source={{ uri: thumbnailPath }} /> : transactionIcon}
          </View>
          <View style={styles.transactionContainer}>
            <View style={styles.transactionRow}>
              <EdgeText style={styles.transactionText}>{transactionText}</EdgeText>
              <EdgeText style={isSentTransaction ? styles.negativeCryptoAmount : styles.positiveCryptoAmount}>{cryptoAmountString}</EdgeText>
            </View>
            <View style={styles.transactionRow}>
              <View style={styles.categoryAndTimeContainer}>
                {categoryText && <EdgeText style={styles.category}>{categoryText}</EdgeText>}
                <EdgeText style={pendingStyle}>{pendingText}</EdgeText>
              </View>
              <EdgeText style={styles.fiatAmount}>{fiatAmountString}</EdgeText>
            </View>
          </View>
        </View>
      </TouchableOpacity>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  rowContainer: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    paddingVertical: theme.rem(1.5),
    paddingHorizontal: theme.rem(1.5)
  },
  iconContainer: {
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: theme.transactionListIconBackground,
    paddingHorizontal: theme.rem(0.5),
    width: theme.rem(3.25),
    height: theme.rem(3.25),
    borderWidth: theme.mediumLineWidth,
    borderRadius: theme.rem(1)
  },
  iconSent: {
    borderColor: theme.negativeText,
    shadowColor: theme.negativeText,
    shadowOffset: {
      width: 0,
      height: theme.rem(3)
    },
    shadowOpacity: 0.5,
    shadowRadius: theme.rem(10),
    elevation: theme.rem(3)
  },
  iconRequest: {
    borderColor: theme.positiveText,
    shadowColor: theme.positiveText,
    shadowOffset: {
      width: 0,
      height: theme.rem(3)
    },
    shadowOpacity: 0.5,
    shadowRadius: theme.rem(10),
    elevation: theme.rem(3)
  },
  icon: {
    width: theme.rem(3),
    height: theme.rem(3),
    borderRadius: theme.rem(0.875)
  },
  transactionContainer: {
    flex: 1,
    justifyContent: 'center',
    paddingHorizontal: theme.rem(0.5)
  },
  transactionRow: {
    flexDirection: 'row',
    marginVertical: theme.rem(0.125)
  },
  transactionText: {
    flex: 1,
    fontFamily: theme.fontFaceBold,
    fontSize: theme.rem(0.75)
  },
  positiveCryptoAmount: {
    fontSize: theme.rem(0.75),
    fontFamily: theme.fontFaceBold,
    color: theme.positiveText,
    textAlign: 'right'
  },
  negativeCryptoAmount: {
    fontSize: theme.rem(0.75),
    fontFamily: theme.fontFaceBold,
    color: theme.negativeText,
    textAlign: 'right'
  },
  fiatAmount: {
    fontSize: theme.rem(0.75),
    color: theme.secondaryText,
    textAlign: 'right'
  },
  categoryAndTimeContainer: {
    flex: 1
  },
  category: {
    fontSize: theme.rem(0.75),
    color: theme.secondaryText
  },
  partialTime: {
    fontSize: theme.rem(0.75),
    color: theme.warningText
  },
  pendingTime: {
    fontSize: theme.rem(0.75),
    color: theme.dangerText
  },
  completedTime: {
    fontSize: theme.rem(0.75),
    color: theme.secondaryText
  }
}))

export const TransactionListRow = connect((state: RootState, ownProps: OwnProps): StateProps => {
  const { currencyCode, walletId, transaction } = ownProps
  const { metadata } = transaction
  const guiWallet = state.ui.wallets.byId[walletId]
  const { fiatCurrencyCode } = guiWallet
  const displayDenomination = getDenomination(currencyCode, state.ui.settings)

  // Required Confirmations
  const { currencyWallets = {} } = state.core.account
  const coreWallet: EdgeCurrencyWallet = currencyWallets[walletId]
  const currencyInfo: EdgeCurrencyInfo = coreWallet.currencyInfo
  const requiredConfirmations = currencyInfo.requiredConfirmations || 1 // set default requiredConfirmations to 1, so once the transaction is in a block consider fully confirmed

  // Thumbnail
  let thumbnailPath
  const contacts = state.contacts || []
  const transactionContactName = metadata && metadata.name ? UTILS.unspacedLowercase(metadata.name) : null
  for (const contact of contacts) {
    const { givenName, familyName } = contact
    const fullName = UTILS.unspacedLowercase(givenName + (familyName || ''))
    if (contact.thumbnailPath && fullName === transactionContactName) {
      thumbnailPath = contact.thumbnailPath
      break
    }
  }

  // CryptoAmount
  const cryptoAmount = UTILS.convertNativeToDisplay(displayDenomination.multiplier)(bns.abs(transaction.nativeAmount || ''))
  const cryptoAmountFormat = intl.formatNumber(UTILS.decimalOrZero(UTILS.truncateDecimals(cryptoAmount, 6), 6))

  // FiatAmount
  const fiatAmount = metadata && metadata.amountFiat ? bns.abs(metadata.amountFiat.toFixed(2)) : '0.00'
  const fiatAmountFormat = intl.formatNumber(bns.toFixed(fiatAmount, 2, 2), { toFixed: 2 })

  return {
    isSentTransaction: UTILS.isSentTransaction(transaction),
    cryptoAmount: cryptoAmountFormat,
    fiatAmount: fiatAmountFormat,
    fiatSymbol: getFiatSymbol(fiatCurrencyCode),
    walletBlockHeight: guiWallet.blockHeight || 0,
    denominationSymbol: displayDenomination.symbol,
    requiredConfirmations,
    selectedCurrencyName: guiWallet.currencyNames[currencyCode] || currencyCode,
    thumbnailPath
  }
})(withTheme(TransactionListRowComponent))
