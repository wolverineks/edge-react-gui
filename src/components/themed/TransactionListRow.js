// @flow

import { bns } from 'biggystring'
import type { EdgeCurrencyInfo, EdgeCurrencyWallet } from 'edge-core-js'
import * as React from 'react'
import { Actions } from 'react-native-router-flux'
import { connect } from 'react-redux'

import * as intl from '../../locales/intl.js'
import s from '../../locales/strings'
import { type RootState } from '../../types/reduxTypes.js'
import type { TransactionListTx } from '../../types/types.js'
import * as UTILS from '../../util/utils'
import { getDenomination, getFiatSymbol } from '../../util/utils.js'
import { showError } from '../services/AirshipInstance.js'
import { TransactionRow } from './TransactionRow.js'

type StateProps = {
  cryptoAmount: string,
  denominationSymbol?: string,
  fiatAmount: string,
  fiatSymbol: string,
  isSentTransaction: boolean,
  requiredConfirmations: number,
  selectedCurrencyName: string,
  thumbnailPath?: string,
  walletBlockHeight: number
}

type OwnProps = {
  walletId: string,
  currencyCode: string,
  transaction: TransactionListTx
}

type Props = OwnProps & StateProps

export class TransactionListRowComponent extends React.PureComponent<Props> {
  onPress = () => {
    const { transaction, thumbnailPath } = this.props
    if (transaction) {
      Actions.transactionDetails({ edgeTransaction: transaction, thumbnailPath })
    } else {
      showError(s.strings.transaction_details_error_invalid)
    }
  }

  render() {
    return (
      <TransactionRow
        cryptoAmount={this.props.cryptoAmount}
        denominationSymbol={this.props.denominationSymbol}
        fiatAmount={this.props.fiatAmount}
        fiatSymbol={this.props.fiatSymbol}
        onPress={this.onPress}
        isSentTransaction={this.props.isSentTransaction}
        requiredConfirmations={this.props.requiredConfirmations}
        selectedCurrencyName={this.props.selectedCurrencyName}
        thumbnailPath={this.props.thumbnailPath}
        transaction={this.props.transaction}
        walletBlockHeight={this.props.walletBlockHeight}
      />
    )
  }
}

export const TransactionListRow = connect((state: RootState, ownProps: OwnProps): StateProps => {
  const { currencyCode, walletId, transaction } = ownProps
  const { metadata } = transaction
  const guiWallet = state.ui.wallets.byId[walletId]
  const { fiatCurrencyCode } = guiWallet
  const displayDenomination = getDenomination(currencyCode, state.ui.settings)

  // Required Confirmations
  const { currencyWallets = {} } = state.core.account
  const coreWallet: EdgeCurrencyWallet = currencyWallets[walletId]
  const currencyInfo: EdgeCurrencyInfo = coreWallet.currencyInfo
  const requiredConfirmations = currencyInfo.requiredConfirmations || 1 // set default requiredConfirmations to 1, so once the transaction is in a block consider fully confirmed

  // Thumbnail
  let thumbnailPath
  const contacts = state.contacts || []
  const transactionContactName = metadata && metadata.name ? UTILS.unspacedLowercase(metadata.name) : null
  for (const contact of contacts) {
    const { givenName, familyName } = contact
    const fullName = UTILS.unspacedLowercase(givenName + (familyName || ''))
    if (contact.thumbnailPath && fullName === transactionContactName) {
      thumbnailPath = contact.thumbnailPath
      break
    }
  }

  // CryptoAmount
  const cryptoAmount = UTILS.convertNativeToDisplay(displayDenomination.multiplier)(bns.abs(transaction.nativeAmount || ''))
  const cryptoAmountFormat = intl.formatNumber(UTILS.decimalOrZero(UTILS.truncateDecimals(cryptoAmount, 6), 6))

  // FiatAmount
  const fiatAmount = metadata && metadata.amountFiat ? bns.abs(metadata.amountFiat.toFixed(2)) : '0.00'
  const fiatAmountFormat = intl.formatNumber(bns.toFixed(fiatAmount, 2, 2), { toFixed: 2 })

  return {
    isSentTransaction: UTILS.isSentTransaction(transaction),
    cryptoAmount: cryptoAmountFormat,
    fiatAmount: fiatAmountFormat,
    fiatSymbol: getFiatSymbol(fiatCurrencyCode),
    walletBlockHeight: guiWallet.blockHeight || 0,
    denominationSymbol: displayDenomination.symbol,
    requiredConfirmations,
    selectedCurrencyName: guiWallet.currencyNames[currencyCode] || currencyCode,
    thumbnailPath
  }
})(TransactionListRowComponent)
