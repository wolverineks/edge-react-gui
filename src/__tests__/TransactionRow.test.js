/* globals describe it expect */
/* eslint-disable flowtype/require-valid-file-annotation */

import { bns } from 'biggystring'
import * as React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'

import { getTheme } from '../components/services/ThemeContext.js'
import { TransactionListRowComponent } from '../components/themed/TransactionListRow.js'
import * as intl from '../locales/intl.js'
import * as UTILS from '../util/utils'

describe('Transaction List Row', () => {
  it('should render props', () => {
    const renderer = new ShallowRenderer()
    const guiWallet = {
      id: 'SXq1f3x21H2e/h5A4ANvrMoK5xs+sQcDoFWHtCG25BA=',
      type: 'wallet:monero',
      name: 'Monero',
      primaryNativeBalance: '1492780012',
      nativeBalances: { XMR: '1492780012' },
      currencyNames: { XMR: 'Monero' },
      currencyCode: 'XMR',
      isoFiatCurrencyCode: 'iso:USD',
      fiatCurrencyCode: 'USD',
      denominations: [{}],
      allDenominations: { XMR: {} },
      metaTokens: [],
      enabledTokens: [],
      receiveAddress: { metadata: {}, nativeAmount: '0', publicAddress: '432hJPUp2C...' },
      blockHeight: 1688551,
      symbolImage: 'https://developer.airbitz.co/content/monero-symbol-orange-64.png',
      symbolImageDarkMono: 'https://developer.airbitz.co/content/monero-symbol-64-87939D.png',
      key: 'SXq1f3x21H2e/h5A4ANvrMoK5xs+sQcDoFWHtCG25BA='
    }
    const transaction = {
      blockHeight: 1683022,
      date: 1539555412.068,
      ourReceiveAddresses: [],
      signedTx: 'no_signature',
      txid: '4e92d23cff1714d52d48c0c5246adf4f6871d6d8d52d774b1b60cc4b28f8f296',
      amountSatoshi: -32295514330000,
      nativeAmount: '-32295514330000',
      networkFee: '0',
      currencyCode: 'XMR',
      wallet: { id: 'SXq1f3x21H…', type: 'wallet:monero' },
      otherParams: {},
      SVGMetadataElement: { name: 'ShapeShift', category: '', notes: 'Exchanged …' },
      dateString: 'Oct 14, 2018',
      time: '3:16 PM',
      key: 0,
      metadata: {
        amountFiat: 4424808418353299.5
      }
    }
    const displayDenomination = {
      multiplier: '1000000000000',
      name: 'XMR',
      symbol: '‎ɱ'
    }
    // CryptoAmount
    const cryptoAmount = UTILS.convertNativeToDisplay(displayDenomination.multiplier)(bns.abs(transaction.nativeAmount || ''))
    const cryptoAmountFormat = intl.formatNumber(UTILS.decimalOrZero(UTILS.truncateDecimals(cryptoAmount, 6), 6))
    // FiatAmount
    const fiatAmount = bns.abs(transaction.metadata.amountFiat.toFixed(2))
    const fiatAmountFormat = intl.formatNumber(bns.toFixed(fiatAmount, 2, 2), { toFixed: 2 })
    const props = {
      cryptoAmount: cryptoAmountFormat,
      denominationSymbol: displayDenomination.symbol,
      fiatAmount: fiatAmountFormat,
      fiatSymbol: UTILS.getFiatSymbol(guiWallet.fiatCurrencyCode),
      isSentTransaction: UTILS.isSentTransaction(transaction),
      requiredConfirmations: 15,
      selectedCurrencyName: guiWallet.currencyNames[guiWallet.currencyCode],
      thumbnailPath: '',
      walletBlockHeight: guiWallet.blockHeight,
      transaction: transaction,
      theme: getTheme()
    }
    const actual = renderer.render(<TransactionListRowComponent {...props} />)

    expect(actual).toMatchSnapshot()
  })
})
