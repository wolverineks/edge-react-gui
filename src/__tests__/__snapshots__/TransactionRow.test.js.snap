// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transaction List Row should render props 1`] = `
<ForwardRef
  onPress={[Function]}
>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginBottom": 3,
        "paddingHorizontal": 34,
        "paddingVertical": 34,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#0C446A",
            "borderRadius": 22,
            "borderWidth": 3,
            "height": 73,
            "justifyContent": "center",
            "paddingHorizontal": 11,
            "width": 73,
          },
          Object {
            "borderColor": "#E85466",
            "elevation": 3,
            "shadowColor": "#E85466",
            "shadowOffset": Object {
              "height": 3,
              "width": 0,
            },
            "shadowOpacity": 0.5,
            "shadowRadius": 10,
          },
        ]
      }
    >
      <Icon
        allowFontScaling={false}
        color="#E85466"
        name="send"
        size={39}
      />
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
          "paddingHorizontal": 11,
        }
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginVertical": 3,
          }
        }
      >
        <WithTheme(EdgeTextComponent)
          style={
            Object {
              "flex": 1,
              "fontFamily": "Quicksand-Medium",
              "fontSize": 17,
            }
          }
        >
          Sent Monero
        </WithTheme(EdgeTextComponent)>
        <WithTheme(EdgeTextComponent)
          style={
            Object {
              "color": "#E85466",
              "fontFamily": "Quicksand-Medium",
              "fontSize": 17,
              "textAlign": "right",
            }
          }
        >
          - ‎ɱ 32.295514
        </WithTheme(EdgeTextComponent)>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginVertical": 3,
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <WithTheme(EdgeTextComponent)
            style={
              Object {
                "color": "#A4C7DF",
                "fontSize": 17,
              }
            }
          >
            3:16 PM
          </WithTheme(EdgeTextComponent)>
        </View>
        <WithTheme(EdgeTextComponent)
          style={
            Object {
              "color": "#A4C7DF",
              "fontSize": 17,
              "textAlign": "right",
            }
          }
        >
          $ 4,424,808,418,353,299.50
        </WithTheme(EdgeTextComponent)>
      </View>
    </View>
  </View>
</ForwardRef>
`;
